gtest/src/wasm_executor.rs

```diff
@@ -0,0 +1,349 @@
+                        externs.push(funcs[key].into())
+                    } else {
+                        panic!("Wasm is asking for unknown function: {:?}. Consider to add in from FuncsHandler", key)
+                    }
+                }
+                _ => continue,
+            };
+        }
+
+        let instance =
+            Instance::new(&mut store, &module, &externs).expect("Failed to create instance");
+        WasmExecutor::set_pages(&mut store, &mut memory, memory_pages)
+            .expect("Failed to set memory pages");
+
+        Self {
+            instance,
+            store,
+            memory,
+            program_id: program.id(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn call_function(&mut self, function_name: &str) -> (Vec<u8>, Vec<JournalNote>) {
+        let function = self.get_function(function_name);
+        let mut prt_to_result_array = [Val::I32(0)];
+
+        function
+            .call(&mut self.store, &[], &mut prt_to_result_array)
+            .expect("Failed call");
+
+        let execution_result = match prt_to_result_array[0] {
+            Val::I32(ptr_to_result) => self.read_result(ptr_to_result),
+            _ => {
+                panic!("{}", "Got wrong type")
+            }
+        };
+        (execution_result, self.get_memory_updates())
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn call_void_function(&mut self, function_name: &str) -> Vec<JournalNote> {
+        let function = self.get_function(function_name);
+        function
+            .call(&mut self.store, &[], &mut [])
+            .expect("Failed call");
+
+        self.get_memory_updates()
+    }
+
+    fn get_memory_updates(&mut self) -> Vec<JournalNote> {
+        let mut pages_data = BTreeMap::new();
+        let mut buffer = [0u8; PageNumber::size()];
+        let memory_size = self.memory.data_size(&mut self.store);
+        for page_number in (0..memory_size).step_by(PageNumber::size()) {
+            if let Err(err) = self.memory.read(&mut self.store, page_number, &mut buffer) {
+                panic!("{:?}", err.to_string())
+            }
+            pages_data.insert(PageNumber::new_from_addr(page_number), buffer.to_vec());
+        }
+        self.build_jornal_notes(pages_data)
+    }
+
+    fn build_jornal_notes(&self, pages_data: BTreeMap<PageNumber, Vec<u8>>) -> Vec<JournalNote> {
+        pages_data
+            .into_iter()
+            .map(|(page_number, data)| JournalNote::UpdatePage {
+                program_id: self.program_id,
+                page_number,
+                data: PageBuf::new_from_vec(data).expect("Failed to convert data to page buffer"),
+            })
+            .collect()
+    }
+
+    fn get_function(&mut self, function_name: &str) -> Func {
+        self.instance
+            .get_func(&mut self.store, function_name)
+            .expect("No function with such name")
+    }
+
+    fn read_result(&mut self, ptr_to_result_data: i32) -> Vec<u8> {
```

Comment:
This low-level part is pretty hard to understand. Going to add some comments here but will that be enough?

Reply:
See below: https://github.com/gear-tech/gear/pull/944#discussion_r892896020